src/Snap/CORS.hs

```diff
@@ -0,0 +1,280 @@
+  , HashableURI(..), HashableMethod (..)
+  ) where
+
+import Control.Applicative
+import Control.Monad (join, when)
+import Data.CaseInsensitive (CI)
+import Data.Hashable (Hashable(..))
+import Data.Maybe (fromMaybe)
+import Data.Text.Encoding (decodeUtf8, encodeUtf8)
+import Network.URI (URI (..), URIAuth (..),  parseURI)
+
+import qualified Data.Attoparsec.Combinator as Attoparsec
+import qualified Data.Attoparsec.ByteString.Char8 as Attoparsec
+import qualified Data.ByteString.Char8 as Char8
+import qualified Data.CaseInsensitive as CI
+import qualified Data.HashSet as HashSet
+import qualified Data.Text as Text
+import qualified Snap.Core as Snap
+
+-- | A set of origins. RFC 6454 specifies that origins are a scheme, host and
+-- port, so the 'OriginSet' wrapper around a 'HashSet.HashSet' ensures that each
+-- 'URI' constists of nothing more than this.
+newtype OriginSet = OriginSet { origins :: HashSet.HashSet HashableURI }
+
+-- | Used to specify the contents of the @Access-Control-Allow-Origin@ header.
+data OriginList
+  = Everywhere
+  -- ^ Allow any origin to access this resource. Corresponds to
+  -- @Access-Control-Allow-Origin: *@
+  | Nowhere
+  -- ^ Do not allow cross-origin requests
+  | Origins OriginSet
+  -- ^ Allow cross-origin requests from these origins.
+
+-- | Specify the options to use when building CORS headers for a response. Most
+-- of these options are 'Snap.Handler' actions to allow you to conditionally
+-- determine the setting of each header.
+data CORSOptions m = CORSOptions
+  { corsAllowOrigin :: m OriginList
+  -- ^ Which origins are allowed to make cross-origin requests.
+
+  , corsAllowCredentials :: m Bool
+  -- ^ Whether or not to allow exposing the response when the omit credentials
+  -- flag is unset.
+
+  , corsExposeHeaders :: m (HashSet.HashSet (CI Char8.ByteString))
+  -- ^ A list of headers that are exposed to clients. This allows clients to
+  -- read the values of these headers, if the response includes them.
+
+  , corsAllowedMethods :: m (HashSet.HashSet HashableMethod)
+  -- ^ A list of request methods that are allowed.
+
+  , corsAllowedHeaders :: HashSet.HashSet String -> m (HashSet.HashSet String)
+  -- ^ An action to determine which of the request headers are allowed.
+  -- This action is supplied the parsed contents of
+  -- @Access-Control-Request-Headers@.
+  }
+
+-- | Liberal default options. Specifies that:
+--
+-- * All origins may make cross-origin requests
+-- * @allow-credentials@ is true.
+-- * No extra headers beyond simple headers are exposed.
+-- * @GET@, @POST@, @PUT@, @DELETE@ and @HEAD@ are all allowed.
+-- * All request headers are allowed.
+--
+-- All options are determined unconditionally.
+defaultOptions :: Monad m => CORSOptions m
+defaultOptions = CORSOptions
+  { corsAllowOrigin = return Everywhere
+  , corsAllowCredentials = return True
+  , corsExposeHeaders = return HashSet.empty
+  , corsAllowedMethods =
+      return $ HashSet.fromList $ map HashableMethod
```

Comment:
Float this set constant out to a top-level CAF to make sure it gets shared. Prefer `return $!` to `return $` (the latter always creates a thunk if the inliner doesn't manage to get rid of it).